### DISCUSSION TITLE
Add support for deno

### DISCUSSION THREAD

**Discussion by Evian-Zhang** [vercel/next.js#26428] [Help] [answered]
**Edit:** See response here: https://github.com/vercel/next.js/discussions/26428#discussioncomment-8283586

**Comment by gustavocastrov** [comment]
Urge!

**Comment by tsujp** [comment]
Any progress on this?

EDIT: I realise my folly, https://github.com/vercel/next.js/issues/12334#issuecomment-633831798

**Comment by timneutkens** [comment]
@tsujp feel free to contribute to making it work.

**Comment by tsujp** [comment]
Planned to have a gander, wanted to chime in here just in-case there are parallel efforts going on.

**Comment by lemarier** [comment]
Hi @timneutkens, I would love to help on this. Something have been planned? 

Did you guys have an idea of what need to be done exactly?

Thanks

**Comment by timneutkens** [comment]
In this case it’ll be entirely up to the people that want to land it. I assume you’ll have to figure out how to run webpack/Babel in deno first though.

**Reply by ild0tt0re** [reply]
Hi @timneutkens! We are still far away? 
After Next.js released in binary format, turbopack towards the stable release and the new routing system (app dir based), deno support would be the last piece of the pie with its unified toolset (lint, test, ...) and would also greatly simplify the maintenance of the project.

**Comment by tsujp** [comment]
I was a bit naive in my "progress on this" above. All the dependencies of Next, and their dependencies' dependencies etc will need to work with deno before Next could -- so I suppose "progress on this" is more of a "progress on deno for production?". I have also heard various people are trying to get dependency auditing (e.g., [deno_udd](https://github.com/hayd/deno-udd)) fleshed out. So these kinds of tools have to be more mature also.

Whoops.

**Comment by webdeb** [comment]
Just created a simple webservice with deno to handle webhooks.
At the end I was able to `deno bundle` the whole application into a **~432 kb** bundle.js file 🎉   
The docker container to run it anywhere was about 66mb. My next.js application container is about 1Gb.

Maybe I am deploying next.js the wrong way, any ideas are welcome

**Comment by kachkaev** [comment]
There is a new project called [Aleph.js](https://alephjs.org), which is

> The React Framework in Deno, inspired by Next.js. 

GitHub: https://github.com/alephjs/aleph.js

Could be worth watching the space to cross-pollinate some ideas 🙂

**Comment by baer** [comment]
At the time of writing, the babel team does not have any plans to support Deno. 

https://github.com/babel/babel/issues/11543#issuecomment-724959695

**Comment by jerrygreen** [comment]
I love how this issue has <kbd>good first issue</kbd> tag, while it’s nearly impossible for NextJS to support Deno. Maybe they can just morally support and say: “Hey guys, gl in whatever you’re doing!” and be done with it, but everyone will consider this 126% rude, so silence is better 😂

Honestly the last time I checked Deno when v1 came out, it was quite poorly working for Windows, and all the feedback I got from their somewhat-core team was almost like: “windows? hah we’re having some kid here not knowing that you have to use Linux to develop things... use WSL” (well, he said somewhat like this), but WSL is almost like little virtual Linux machine, so mehhh... While NodeJS is working without any virtual machine layer completely fine, it’s a shame Deno can’t be run that easily too. So despite Deno v1 came out, they still have a lot of work to do to improve their fundamental things like cross-platform support (at the very least)

Their core-core member Ryan Dahl sympathizes me actually: he has good vision, good ideas, so maybe some Deno v2 will be much better and more production-ready, who knows...

_💕waves tons of support 💕_

**Reply by jerrygreen** [reply]
For those who dislike the message, a question:

What are you expecting when you say: «Add support for Deno in NextJS»?

If you mean the ability to create a NextJS project within Deno, then it is not possible without rewriting the whole framework from scratch: all the js dependencies should be removed or replaced with deno equivalents, most of the glue-code should be rewritten also, leaving only the handful logic that can remain unchanged. This is not «Add support». This is a complete change of plans, complete another project like AlephJS. NextJS cannot help here except literally to wish them luck.

**Comment by shadowtime2000** [comment]
@JerryGreen It seems your points about Deno needing to improve cross-platform support is based on Deno v1 which was released around a year ago. I have been using Deno and plain old Windows without WSL for AlephJS and other projects, and it has been working great. You can't really base it's flaws around the v1 release which was around a year ago. I recommend you download the latest Deno, v1.8.2, and try it out because those "fundamental things" have been worked on.

**Comment by jerrygreen** [comment]
@shadowtime2000 time flows fast 😅 Thank you for saying all that, I’ll keep in mind Deno deserves another try.

I saw AlephJS back in the days too, and yeah: if you wanna use Deno and want NextJS then AlephJS is definitely your choice, despite having some differences it fits the most.

For some other things, like Electron, people do have equivalent questions though:

https://github.com/electron/electron/issues/23613

Unfortunately there’s no such thing as Denotron or something 😞 Nor there’s support from Electron (neither currently nor in plans). If there would be, I would be probably making my projects with Deno already

**Reply by DanielBailey-web** [reply]
Tauri is trying to compete in the same space as electron except they have much smaller bundle sizes. 

They are planning to support deno (hopefully soon). You can currently make next.js app using it already. 

I think its an interesting piece of software.
https://tauri.studio/

Can only target Windows, Linux, and Mac currently (no ios/android) (ios is in progress)
https://github.com/tauri-apps/tauri#platforms

**Reply by jerrygreen** [reply]
@DanielBailey-web yeah I’ve seen Tauri maybe like a year ago, tried some hello-world projects with that, and it looks interesting. Unfortunately I haven’t had enough time, opportunity, and desire, - to try it in some more combat-like real-worldish projects. But this instrument looks real interesting 🧐

**Comment by soupCodez** [comment]
Deno is adding support for npm so that's gonna be pretty interesting

**Reply by jerrygreen** [reply]
Have you already tried NextJS from within Deno? I think the thing would be pretty fucked up, isn’t it? Just curious 🧐

**Comment by damienmillet** [comment]
It’s a good question !

**Comment by TwilightSublime** [comment]
Has anyone tried running "deno run -A npm:create-next-app" ?

**Comment by TwilightSublime** [comment]
![image](https://user-images.githubusercontent.com/84061672/204088988-87bcc229-6e7a-41c5-a549-4c7270142498.png)
it works but it's stuck at download stage

**Reply by TwilightSublime** [reply]
Tested on deno 1.28.2 btw

**Comment by officialrajdeepsingh** [comment]
Finally, find something interesting @leerob

**Comment by TwilightSublime** [comment]
Looks like this needs to be done first
https://github.com/denoland/deno/issues/16164

**Comment by iugo** [comment]
In the Deno community, there is an issue discussing this matter. https://github.com/denoland/deno/issues/16679

**Comment by TwilightSublime** [comment]
update! it works

run `deno run -A npm:create-next-app` in the terminal

**Reply by soupCodez** [reply]
Amazing!

**Reply by alexgleason** [reply]
Are you able to actually run a next app after that? Or do you get stuck at another point?

**Reply by officialrajdeepsingh** [reply]
`deno run -A npm:create-next-app` command is working, but it also create `node_modules` folder. when we  running `deno task dev` command. it is not working.
It is not working

**Reply by ghost** [reply]
@officialrajdeepsingh you'll have to edit the package.json script to use deno

**Reply by philip-taiwan** [reply]
Could you show your config? Thanks

**Reply by ShinobiWPS** [reply]
any update on this?

**Reply by lpknv** [reply]
I tried running `deno run -A npm:create-next-app` and it is still working. Running `deno task dev` resulted in this error:

```
Task dev next dev
error: Uncaught (in promise) ReferenceError: module is not defined
module.exports = nextConfig
^
    at file:///home/nik/code/deno-next-app/next.config.js:4:1
```

Trying to change my next.config.js (out of curiosity) to
```
/** @type {import('next').NextConfig} */
const nextConfig = {};

**export default nextConfig;**
```

and running `deno task dev` gets me stuck in downloading `zod`:

```
Task dev next dev
Download https://registry.npmjs.org/zod (47/51)
```

**Comment by Allan-Nava** [comment]
Any news? +1

**Comment by omnaiduu** [comment]
Do we actually want to support to deploy to deno or actually use or create project in deno.. i am confused.. for the first problem.. we can do it..

**Comment by GuilhermeWerner** [comment]
A use case that would be useful for me would be to deploy a website with nextjs with deno, for desktop, without having to have nodejs as a dependency.

**Comment by orzklv** [comment]
Any news? +1

**Comment by MobinX** [comment]
?

**Comment by ShinobiWPS** [comment]
Well, since this is not an Issue we can talk with no reply for years to come :)

**Comment by alexgleason** [comment]
It's possible to run other bundlers like Vite with Deno now, but that doesn't mean it's good. What I actually want is the ability to import code by URL in my client-side JS, as well as use Deno's other great features like the built-in formatter, linter, and testing framework.

But Vite, and probably Next.js, expects you to import code from `node_modules`. Even if you manage to get the program running in Deno, Deno has no `package.json`. So you have to resort to importing code in a `deps.ts` type of file, and then VSCode does not work at all, even with the Deno extension. Everything is underlined in red and type checking doesn't work. This is the current experience with Deno + Vite.

So what does it mean to "add support for Deno"? I think what people want is for the bundler to become aware of all the things Deno can do, and support them. But it's beyond just adding a polyfill here or there to get it to run. What I'm saying is that even AFTER it runs doesn't mean it will be good or even worth using.

That being said, I do believe Deno is the future of server-side JS, and I hope Vercel finds a way to step into the future.

**Comment by leerob** [comment, marked as answer]
**Edit:** Good news, Deno 2 now supports Next.js!

https://deno.com/blog/v2.0
